### PR TITLE
Improved types (PHPStan level 6)

### PR DESCRIPTION
**Pull request by aivchen**

### CODE REVIEW

File: lib/Model/Benchmark.php [hunk 1]
```diff
@@ -110,7 +106,7 @@ public function getSuite(): Suite
     }
 
     /**
-     * @return ArrayIterator<Subject>
+     * @return ArrayIterator<array-key, Subject>
```
Comment (dantleech) [inline lib/Model/Benchmark.php:109]: is the `ArrayIterator<Subject>` form not valid? but happy with the more explict form in anycase

File: lib/Model/Error.php [hunk 2]
```diff
@@ -20,6 +20,14 @@
...
 {
+    /**
+     * @param string $message
+     * @param string $class
+     * @param int $code
+     * @param string $file
+     * @param int $line
+     * @param string $trace
+     */
     public function __construct(private $message, private $class, private $code, private $file, private $line, private $trace)
```
Comment (dantleech) [inline lib/Model/Error.php:31]: any reason not to add real type hints here?

File: lib/Model/Error.php [hunk 3]
```diff
@@ -36,31 +44,49 @@ public static function fromException(Throwable $exception): Error
...
+     */
     public function getClass()
     {
         return $this->class;
     }
 
+    /**
+     * @return string
+     */
     public function getTrace()
```
Comment (dantleech) [inline lib/Model/Error.php:90]: ditto, seems these can all be real return types?

File: lib/Model/ResultCollection.php [hunk 4]
```diff
@@ -106,7 +113,12 @@ public function getMetric(string $class, string $metric)
...
      * clearly a problem and we should allow an error to be thrown.
      *
+     * @template TDefault
+     *
+     * @param class-string<ResultInterface> $class
+     * @param TDefault $default
      *
+     * @return int|float|TDefault
      */
     public function getMetricOrDefault(string $class, string $metric, $default = null)
```
Comment (dantleech) [inline lib/Model/ResultCollection.php:123]: :+1:

File: lib/Model/Summary.php [hunk 5]
```diff
@@ -26,6 +26,8 @@ class Summary
     private int $nbFailures = 0;
     private int $nbAssertions = 0;
     private int $nbErrors = 0;
+
+    /** @var array<string, array<int|float>> */
```
Comment (dantleech) [inline lib/Model/Summary.php:30]: _could_ use an array shape here, although maybe better as a type alias if it's referenced again

File: lib/Model/Summary.php [hunk 6]
```diff
@@ -120,41 +122,65 @@ public function getNbAssertions(): int
         return $this->nbAssertions;
     }
 
+    /**
+     * @return array<string, array<int|float>>
+     */
     public function getStats(): array
```
Comment (dantleech) [inline lib/Model/Summary.php:128]: i.e. we could return that array shape here

File: lib/Model/Summary.php [hunk 7]
```diff
@@ -120,41 +122,65 @@ public function getNbAssertions(): int
...
+     */
     public function getMeanStDev()
     {
         return Statistics::mean($this->stats['stdev']);
     }
 
+    /**
+     * @return int|float
+     */
     public function getMeanRelStDev()
```
Comment (dantleech) [inline lib/Model/Summary.php:184]: could use real RTs?

File: lib/Model/Variant.php [hunk 8]
```diff
@@ -94,6 +93,7 @@ public function createIteration(array $results = []): Iteration
     /**
      * Return the iteration at the given index.
      *
+     * @param int $index
      */
     public function getIteration($index): ?Iteration
```
Comment (dantleech) [inline lib/Model/Variant.php:98]: ...

**Review by dantleech — COMMENTED** [review]
nice work, I guess you opted to use docblocks rather than real types for B/C reasons? i wonder if we can add the real types and either:

1) bump to 2.0
2) assume that nobody has overridden this stuff (it's highly unlikely)
3) keep the docblocks and migrate them at a later date with rector.

happy to got with 3)

File: lib/Model/Benchmark.php (hunk 1 above)
Comment (aivchen) [reply lib/Model/Benchmark.php:109]: Yes, it's not valid:

> Method PhpBench\Model\Benchmark::getIterator() return type with generic class ArrayIterator does not specify its types: TKey, TValue

File: lib/Model/Error.php (hunk 2 above)
Comment (aivchen) [reply lib/Model/Error.php:31]: B/C

File: lib/Model/Error.php (hunk 3 above)
Comment (aivchen) [reply lib/Model/Error.php:90]: B/C

File: lib/Model/Summary.php (hunk 7 above)
Comment (aivchen) [reply lib/Model/Summary.php:184]: B/C :smile:

File: lib/Model/Variant.php (hunk 8 above)
Comment (aivchen) [reply lib/Model/Variant.php:98]: B/C

**Comment by aivchen** [comment]
> I guess you opted to use docblocks rather than real types for B/C reasons?

Exactly:)

>2. assume that nobody has overridden this stuff (it's highly unlikely)

It's unlikely but we can't exclude it:)

> 3. keep the docblocks and migrate them at a later date with rector.

Yes. I think we need to type eveything with dockblocks for now. After that we can prepare 2.0 and run rector. Also will be cool to finalize all classes in 2.0 to have more freedom in classes changing.

File: lib/Model/Summary.php (hunk 5 above)
Comment (aivchen) [reply lib/Model/Summary.php:30]: Let me think a bit.

**Comment by dantleech** [comment]
Sounds good!

**Pull request merged**